BigQuery/src/BigQueryClient.php

```diff
@@ -751,4 +753,158 @@ public function getServiceAccount(array $options = [])
         $resp = $this->connection->getServiceAccount($options + ['projectId' => $this->projectId]);
         return $resp['email'];
     }
+
+    /**
+     * Returns a copy job configuration to be passed to either
+     * {@see Google\Cloud\BigQuery\BigQueryClient::runJob()} or
+     * {@see Google\Cloud\BigQuery\BigQueryClient::startJob()}. A
+     * configuration can be built using fluent setters or by providing a full
+     * set of options at once.
+     *
+     * Example:
+     * ```
+     * $sourceTable = $bigQuery->dataset('samples', 'bigquery-public-data')
+     *     ->table('shakespeare');
+     * $destinationTable = $bigQuery->dataset('myDataset')
+     *     ->table('myDestinationTable');
+     *
+     * $copyJobConfig = $bigQuery->copyTable($sourceTable, $destinationTable);
+     * ```
+     *
+     * @see https://cloud.google.com/bigquery/docs/reference/v2/jobs Jobs insert API Documentation.
+     *
+     * @param Table $source The source table.
+     * @param Table $destination The destination table.
+     * @param array $options [optional] Please see the
+     *        [upstream API documentation for Job configuration]
+     *        (https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration)
+     *        for the available options.
+     * @return CopyJobConfiguration
+     */
+    public function copyTable(Table $source, Table $destination, array $options = [])
+    {
+        return (new CopyJobConfiguration(
+            $this->projectId,
+            $options,
+            $this->location
+        ))
+            ->sourceTable($source)
+            ->destinationTable($destination);
+    }
+
+    /**
+     * Returns an extract job configuration to be passed to either
+     * {@see Google\Cloud\BigQuery\BigQueryClient::runJob()} or
+     * {@see Google\Cloud\BigQuery\BigQueryClient::startJob()}. A
+     * configuration can be built using fluent setters or by providing a full
+     * set of options at once.
+     *
+     * Example:
+     * ```
+     * $destinationObject = 'gs://myBucket/tableOutput';
+     * $extractJobConfig = $bigQuery->extractTable($table, $destinationObject);
+     * ```
+     *
+     * @see https://cloud.google.com/bigquery/docs/reference/v2/jobs Jobs insert API Documentation.
+     *
+     * @param Table $source The source table.
+     * @param string|StorageObject $destination The destination object. May be
+     *        a {@see Google\Cloud\Storage\StorageObject} or a URI pointing to
+     *        a Google Cloud Storage object in the format of
+     *        `gs://{bucket-name}/{object-name}`.
+     * @param array $options [optional] Please see the
+     *        [upstream API documentation for Job configuration]
+     *        (https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#configuration)
+     *        for the available options.
+     * @return ExtractJobConfiguration
+     */
+    public function extractTable(Table $source, $destination, array $options = [])
```

Comment:
Can you explain in more depth (maybe with examples) why these are required and we cannot make do with the existing methods?

Reply:
Well, _technically_, we can use existing methods.

Job configuration contains id of instance which will run that job. This id is passed to job configuration constructor, and there is no nice and easy way to change it. And methods like `Table#extract()` pass id of instance this table belongs to. So if we want to run a job on one instance that uses a table from another one, we need to use some ugly tricks to make a proper job configuration.

Methods like `BigQueryClient#extractTable()` make such tasks less ugly.

Reply:
Does `instance` refer to the projectId? Would it be possible to add projectId to the `$options` array of each job config factory?

Alternatively, perhaps following the pattern established by the addition of the start/run job methods in the other pull request, job config factories would be ok on `BigQueryClient`. If so, I'd like them to be just `copy`, `extract`, `load`. The user should be responsible for configuring it on a more granular level. That will future-proof us against the addition of jobs against different resource types, preventing the problem the other PR solves from happening again.

If users want the additional convenience of the table passing itself into the job config, they can still use the existing factory method on `Table`, and so forth.

I'd love to have @dwsupplee weigh in on this as well, but if this makes sense to you we can move ahead with it.

Reply:
> Does instance refer to the projectId?

Yes, just a typo.

> Would it be possible to add projectId to the `$options` array of each job config factory?

Yes, it's possible. For now client can provide project id twice: as `projectId` and as `jobReference.projectId` options, and these two must match. (Client should also note `jobReference.location` option). Perhaps we should cut off a corner here and set `jobReference.projectId` equal to `projectId` automatically.

> If so, I'd like them to be just `copy`, `extract`, `load`. The user should be responsible for configuring it on a more granular level. That will future-proof us against the addition of jobs against different resource types, preventing the problem the other PR solves from happening again.

Do you mean that these methods should return "bare" job configurations, and then clients should call methods like `sourceTable()`/`sourceModel()`, `destinationUris()`? Yes, this makes sense.